src/core/refiner.ml

```diff
@@ -0,0 +1,316 @@
+        [eq_modulo]. *)
+       begin
+         if !Debug.log_enabled then
+           log (Extra.yel "add constraint %a") Print.pp_constr
+             (ctx, a, b);
+         Stdlib.(constraints := (ctx, a, b) :: !constraints)
+       end
+
+    (** [coerce ctx t a b] refines term [t] of type [a] into term of type
+        [b]. *)
+    let rec coerce : ctxt -> term -> term -> term -> term =
+     fun ctx t a b ->
+      if !Debug.log_enabled then
+        log "Coerce [%a: %a = %a]" Print.pp_term t Print.pp_term a Print.pp_term
+          b;
+      if Eval.eq_modulo ctx a b then t
+      else match L.lookup ctx a b with
+        | Some (c, eqs) ->
+            let add_eq (ctx, a, b) = unif ctx a b in
+            List.iter add_eq eqs;
+            let out = Bindlib.subst c t in
+            if !(Debug.log_enabled) then
+              log (Extra.gre "Coerced \"%a\" to \"%a\"") Print.pp_term t
+                Print.pp_term out;
+            out
+        | None ->
+            (if !Debug.log_enabled then log (Extra.red "Failed coercion"));
+            unif ctx a b; t
+
+    (** [type_enforce ctx a] refines type [a] in context [ctx]. [a] must be a
+        type. *)
+    and type_enforce : ctxt -> term -> term * term =
+     fun ctx a ->
+      if !Debug.log_enabled then log "Type enforce [%a]" Print.pp_term a;
+      let a, s = infer ctx a in
+      let sort =
+        match unfold s with
+        | Kind -> Kind
+        | Type -> Type
+        | _ -> Type
+        (* FIXME The algorithm should be able to backtrack on the choice of
+           this sort, first trying [Type], and if it does not succeed, trying
+           [Kind]. *)
+      in
+      let a = coerce ctx a s sort in
+      (a, s)
+
+    (** [force ctx t a] returns refined term [t] which has type [a] expected by
+        its context [ctx]. *)
+    and force : ctxt -> term -> term -> term =
+     fun ctx te ty ->
+      if !Debug.log_enabled then
+        log "Force [%a] of [%a]" Print.pp_term te Print.pp_term ty;
+      let default () =
+        let t', a' = infer ctx te in
+        coerce ctx t' a' ty
+      in
+      match unfold te with
+      | Abst (dom, t) -> (
+          match Eval.whnf ctx ty with
+          | Prod (e1, e2) ->
+              let dom, _ = type_enforce ctx dom in
+              unif ctx dom e1;
+              let x, t, e2 = Bindlib.unbind2 t e2 in
+              let ctx = (x, dom, None) :: ctx in
+              let t = force ctx t e2 in
+              Abst (dom, Bindlib.(lift t |> bind_var x |> unbox))
+          | _ -> default () )
+      | LLet (t_ty, t, u) ->
+          let t_ty, _ = type_enforce ctx t_ty in
+          let t = force ctx t t_ty in
+          let x, u, ctx = Ctxt.unbind ctx t_ty (Some t) u in
+          let ty = Bindlib.(lift ty |> bind_var x |> unbox) in
+          let ty = Bindlib.subst ty t in
+          let u = force ctx u ty in
+          let u = Bindlib.(lift u |> bind_var x |> unbox) in
+          LLet (t_ty, t, u)
```

Comment:
What is the advantage of having special cases for Abst and LLet? Because it looks like code duplication. Is it more efficient in practice?

Reply:
I tested your branch and, actually, this is slower: make tests takes 11.863s with no special cases instead of 13.861s with special cases. So, it seems better not to have special cases.

Reply:
However these special cases allow to infer more concise type, and take advantage of the user defined types. Since we intend to have an interactive tool, it makes sense to keep them, in my opinion. Perhaps we want a "batch" mode that would skip those cases?

Reply:
And in some cases (an example would be welcome), it may generate simpler unification problems.

Reply:
Perhaps, but a time increase of 17% seems really too much.

Reply:
I found that these cases are required to insert coercions into the domain of abstractions, and into the type annotation of local lets.

Reply:
Nope, my mistake, coercions are inserted thanks to
```ocaml
      | Abst (dom, b) ->
          let dom, _ = type_enforce ctx dom in
```
in `infer`